4.25 unconditionally uses `uselocale()`

### Welcome

- [x] Yes, I'm using the latest major release or the current development version. These are the only supported versions.
- [x] Yes, I've searched similar issues and discussions on GitHub and didn't find any.

### Current Behavior

I'm trying to update i3wm in pkgsrc to 4.25, but the build fails with:
```
../src/ipc.c: In function 'handle_tree':
../src/ipc.c:912:28: error: implicit declaration of function 'uselocale'; did you mean 'setlocale'? [-Wimplicit-function-declaration]
  912 |     locale_t prev_locale = uselocale(numericC);
      |                            ^~~~~~~~~
      |                            setlocale
```
and many similar lines.

NetBSD does not provide `uselocale`.

### Expected Behavior

It would be good if the build checked for `uselocale` and only used it if the system provides it.

### Reproduction Instructions

Try building on NetBSD, in my case NetBSD 11.99.4/x86_64.

### i3 version

```text
From the last successfully building version:

Binary i3 version:  4.24 (2024-11-06) © 2009 Michael Stapelberg and contributors

but 4.25 doesn't build.
```

### Config file

```text
n/a
```

### Linux distribution & Version

pkgsrc on NetBSD

### Are you using a compositor?

Other

### Logfile

_No response_

--- a/include/main.h
+++ b/include/main.h
@@ -10,9 +10,6 @@
 #pragma once
 
 #include <config.h>
-#include <locale.h>
-
-extern locale_t numericC;
 
 /**
  * Enable or disable the main X11 event handling function.

--- a/src/ipc.c
+++ b/src/ipc.c
@@ -909,10 +909,12 @@ static void dump_bar_config(yajl_gen gen, Barconfig *config) {
 }
 
 IPC_HANDLER(tree) {
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     yajl_gen gen = ygenalloc();
     dump_node(gen, croot, false);
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 
     const unsigned char *payload;
     ylength length;
@@ -1585,7 +1587,8 @@ ipc_client *ipc_new_client_on_fd(EV_P_ int fd) {
  * generator. Free with yajl_gen_free().
  */
 yajl_gen ipc_marshal_workspace_event(const char *change, Con *current, Con *old) {
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     yajl_gen gen = ygenalloc();
 
     y(map_open);
@@ -1609,7 +1612,8 @@ yajl_gen ipc_marshal_workspace_event(const char *change, Con *current, Con *old)
 
     y(map_close);
 
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 
     return gen;
 }
@@ -1639,7 +1643,8 @@ void ipc_send_window_event(const char *property, Con *con) {
     DLOG("Issue IPC window %s event (con = %p, window = 0x%08x)\n",
          property, con, (con->window ? con->window->id : XCB_WINDOW_NONE));
 
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     yajl_gen gen = ygenalloc();
 
     y(map_open);
@@ -1658,7 +1663,8 @@ void ipc_send_window_event(const char *property, Con *con) {
 
     ipc_send_event("window", I3_IPC_EVENT_WINDOW, (const char *)payload);
     y(free);
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 }
 
 /*
@@ -1666,7 +1672,8 @@ void ipc_send_window_event(const char *property, Con *con) {
  */
 void ipc_send_barconfig_update_event(Barconfig *barconfig) {
     DLOG("Issue barconfig_update event for id = %s\n", barconfig->id);
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     yajl_gen gen = ygenalloc();
 
     dump_bar_config(gen, barconfig);
@@ -1677,7 +1684,8 @@ void ipc_send_barconfig_update_event(Barconfig *barconfig) {
 
     ipc_send_event("barconfig_update", I3_IPC_EVENT_BARCONFIG_UPDATE, (const char *)payload);
     y(free);
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 }
 
 /*
@@ -1686,7 +1694,8 @@ void ipc_send_barconfig_update_event(Barconfig *barconfig) {
 void ipc_send_binding_event(const char *event_type, Binding *bind, const char *modename) {
     DLOG("Issue IPC binding %s event (sym = %s, code = %d)\n", event_type, bind->symbol, bind->keycode);
 
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
 
     yajl_gen gen = ygenalloc();
 
@@ -1714,7 +1723,8 @@ void ipc_send_binding_event(const char *event_type, Binding *bind, const char *m
     ipc_send_event("binding", I3_IPC_EVENT_BINDING, (const char *)payload);
 
     y(free);
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 }
 
 /*

--- a/src/load_layout.c
+++ b/src/load_layout.c
@@ -634,14 +634,16 @@ bool json_validate(const char *buf, const size_t len) {
     /* Allow multiple values, i.e. multiple nodes to attach */
     yajl_config(hand, yajl_allow_multiple_values, true);
 
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     if (yajl_parse(hand, (const unsigned char *)buf, len) != yajl_status_ok) {
         unsigned char *str = yajl_get_error(hand, 1, (const unsigned char *)buf, len);
         ELOG("JSON parsing error: %s\n", str);
         yajl_free_error(hand, str);
         valid = false;
     }
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 
     yajl_complete_parse(hand);
     yajl_free(hand);
@@ -671,7 +673,8 @@ json_content_t json_determine_content(const char *buf, const size_t len) {
     yajl_config(hand, yajl_allow_comments, true);
     /* Allow multiple values, i.e. multiple nodes to attach */
     yajl_config(hand, yajl_allow_multiple_values, true);
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     const yajl_status stat = yajl_parse(hand, (const unsigned char *)buf, len);
     if (stat != yajl_status_ok && stat != yajl_status_client_canceled) {
         unsigned char *str = yajl_get_error(hand, 1, (const unsigned char *)buf, len);
@@ -679,7 +682,8 @@ json_content_t json_determine_content(const char *buf, const size_t len) {
         yajl_free_error(hand, str);
     }
 
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
     yajl_complete_parse(hand);
     yajl_free(hand);
 
@@ -725,7 +729,8 @@ void tree_append_json(Con *con, const char *buf, const size_t len, char **errorm
     parsing_geometry = false;
     parsing_focus = false;
     parsing_marks = false;
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     const yajl_status stat = yajl_parse(hand, (const unsigned char *)buf, len);
     if (stat != yajl_status_ok) {
         unsigned char *str = yajl_get_error(hand, 1, (const unsigned char *)buf, len);
@@ -750,7 +755,8 @@ void tree_append_json(Con *con, const char *buf, const size_t len, char **errorm
      * next time. */
     con_fix_percent(con);
 
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
     yajl_complete_parse(hand);
     yajl_free(hand);
 

--- a/src/main.c
+++ b/src/main.c
@@ -36,8 +36,6 @@
 #include "i3-atoms_NET_SUPPORTED.xmacro.h"
 #include "i3-atoms_rest.xmacro.h"
 
-locale_t numericC;
-
 /* The original value of RLIMIT_CORE when i3 was started. We need to restore
  * this before starting any other process, since we set RLIMIT_CORE to
  * RLIM_INFINITY for i3 debugging versions. */
@@ -316,7 +314,6 @@ int main(int argc, char *argv[]) {
     int option_index = 0, opt;
 
     setlocale(LC_ALL, "");
-    numericC = newlocale(LC_NUMERIC_MASK, "C", 0);
 
     /* Get the RLIMIT_CORE limit at startup time to restore this before
      * starting processes. */

--- a/src/util.c
+++ b/src/util.c
@@ -221,12 +221,14 @@ static char **add_argument(char **original, char *opt_char, char *opt_arg, char 
 #define ystr(str) yajl_gen_string(gen, (unsigned char *)str, strlen(str))
 
 static char *store_restart_layout(void) {
-    locale_t prev_locale = uselocale(numericC);
+    char *prev_locale = sstrdup(setlocale(LC_NUMERIC, NULL));
+    setlocale(LC_NUMERIC, "C");
     yajl_gen gen = yajl_gen_alloc(NULL);
 
     dump_node(gen, croot, true);
 
-    uselocale(prev_locale);
+    setlocale(LC_NUMERIC, prev_locale);
+    free(prev_locale);
 
     const unsigned char *payload;
     size_t length;

